gamemodes/terrortown/entities/weapons/weapon_zm_carry.lua

```diff
@@ -356,100 +294,123 @@ end
-		and (target:GetClass() ~= "prop_ragdoll" or GetGlobalBool("ttt_ragdoll_carrying"))
-		and (not target:IsWeapon() or GetGlobalBool("ttt_weapon_carrying"))
+		and (ctype ~= CARRY_TYPE_RAGDOLL or cvAllowRagCarry:GetBool())
+		and (ctype ~= CARRY_TYPE_WEAPON or cvAllowWepCarry:GetBool())
 		---
 		-- @realm shared
 		and not hook.Run("TTT2PlayerPreventPickupEnt", ply, target)
 end
 
+---
+-- @param Entity target
+-- @realm shared
+function SWEP:AllowPush(target)
+	local phys = target:GetPhysicsObject()
+
+	if not IsValid(phys)
+		or not phys:IsMoveable()
+		or phys:HasGameFlag(FVPHYSICS_PLAYER_HELD)
+	then
+		 return false
+	end
+
+	return true
+end
+
+---
+-- @param Entity trEnt
+-- @param TraceResult trace
+-- @param boolean pull Whether to pull it, instead of push
+-- @realm shared
+function SWEP:PushObject(trEnt, trace, pull)
+	-- pull heavy stuff
+	local phys2 = trEnt:GetPhysicsObject()
+	local pdir = trace.Normal
+	if pull then
+		pdir = pdir * -1
+	end
+
+	local is_ragdoll = DetermineCarryType(trEnt) == CARRY_TYPE_RAGDOLL
+	if is_ragdoll then
+		phys2 = trEnt:GetPhysicsObjectNum(trace.PhysicsBone)
+		-- increase refire to make rags easier to drag
+		--self.Weapon:SetNextSecondaryFire(CurTime() + 0.04)
+	end
+
+	if IsValid(phys2) then
+		self:MoveObject(phys2, pdir, self.moveForce, is_ragdoll)
+
+		return true
+	end
+end
+
 ---
 -- @param boolean pickup
 -- @realm shared
 function SWEP:DoAttack(pickup)
 	self:SetNextPrimaryFire(CurTime() + self.Primary.Delay)
 	self:SetNextSecondaryFire(CurTime() + self.Secondary.Delay)
 
-	if IsValid(self.EntHolding) then
+	-- allows us to "drop" on SecondaryAttack, and "release" with PrimaryAttack
+	local want_throw = cvPropThrow:GetBool() and not pickup
+
+	local ctarget = self:GetCarryTarget()
+	local ctype = self:GetCarryType()
+	if IsValid(ctarget) then
 		self:SendWeaponAnim(ACT_VM_MISSCENTER)
 
-		if not pickup and self.EntHolding:GetClass() == "prop_ragdoll" then
+		if not pickup and ctype == CARRY_TYPE_RAGDOLL then
 			-- see if we can pin this ragdoll to a wall in front of us
 			if not self:PinRagdoll() then
 				-- else just drop it as usual
-				self:Drop()
+				self:Drop(false)
 			end
 		else
-			self:Drop()
+			self:Drop(want_throw)
 		end
 
 		self:SetNextSecondaryFire(CurTime() + 0.3)
 
 		return
 	end
 
+	-- if we let the client mess with physics, desync ensues
+	if CLIENT then return end
+
 	local ply = self:GetOwner()
-	local trace = ply:GetEyeTrace(MASK_SHOT)
+	local trace = ply:GetEyeTrace()
 	local trEnt = trace.Entity
 
-	if IsValid(trEnt) then
-		local phys = trEnt:GetPhysicsObject()
+	if not IsValid(trEnt) then return end
+	if not self:AllowPush(trEnt) then return end
 
-		if not IsValid(phys) or not phys:IsMoveable() or phys:HasGameFlag(FVPHYSICS_PLAYER_HELD) then return end
-
-		-- if we let the client mess with physics, desync ensues
-		if CLIENT then return end
+	local within_range = (ply:EyePos() - trace.HitPos):Length() < self:GetRange(trEnt)
 
+	if within_range then
```

Comment:
this one could be converted to a `if not then return end` to prevent deep scope pyramids